docs/source/plugin/what.rst

```diff
@@ -0,0 +1,133 @@
+=================
+What is a Plugin?
+=================
+
+Before writing your first plugin, you may want to know what is a plugin, how is
+it made, and what are the best solutions for your need. In this chapter we'll
+cover what is a Plugin, what you can do with them, and some vocabulary.
+
+You may want to skip ahead to :doc:`anatomy` if you already know what a Plugin
+is and what types of plugin there are.
+
+.. contents::
+    :local:
+    :depth: 2
+
+Plugin Types
+============
+
+There are four types of plugins: :term:`Single file plugin`,
+:term:`Folder plugin`, :term:`Namespace package plugin`, and
+:term:`Entry point plugin`.
+
+Single file
+-----------
+
+A **Single file** plugin is the most basic form a Sopel plugin can take. It is
+composed of a single file available to Sopel from its plugin's directory (or
+from one of its
+:attr:`extra directories <sopel.config.core_section.CoreSection.extra>`). You
+should write a single file plugin when:
+
+* everything is in a single file
+* you don't have any dependencies other than Sopel and the Python's built-in
+  library
+
+This type of plugin is easy to install and share: copy its file into the right
+location, and Sopel will load it on startup.
+
+Folder
+------
+
+A **Folder** plugin is a folder that contains a ``__init__.py`` file. It is
+available to Sopel the same way a :ref:`Single file` plugin is, and share the
+same limitation. Everything that is not available in this ``__init__.py`` file
+won't be visible to Sopel.
+
+By default, Sopel's plugin directories are **not** available in ``sys.path``,
+so it is not possible to do ``from my_plugin_dir import some_module`` without
+a specific configuration from the bot's owner.
+
+When a Plugin author wants to split code into different files, and/or share it
+between multiple plugins, the best option is to use an :ref:`Entry point`
+plugin instead.
+
+Namespace package
+-----------------
+
+A **Namespace package** plugin is a Python :term:`namespace package`, using
+the namespace ``sopel_modules``. It must installed in the Python's environement
+to be used by Sopel, and can require Python dependencies.
+
+Given a ``sopel_modules.plugin`` plugin, Sopel will load everything that is
+available from ``sopel_modules/plugin/__init__.py`` file.
+
+It is the initial version of Sopel's packaged plugins: it can be packaged and
+uploaded to `PyPI`_ then installed using ``pip install``.
+
+When a Plugin author wants to distribute a Sopel plugin, the best option is to
+use an :ref:`Entry point` plugin instead.
```

Comment:
May be worth mentioning older Sopel version compatibility here and the other places entry-point plugins get recommended, since <7 doesn't support entry points.

Reply:
I'll add a note in the `Entry point` section, but I'm not fond of adding it to all the other places. After all, this is the last up-to-date version of the documentation, Sopel 7 has been out for a while now, and this won't be published online until 7.1 comes out. Beside, future version of Sopel going forward will support entry points.

So, agree if it was an archive/a history book, disagree with a documentation that moves forward.

Reply:
Ah, true. I guess we can pretty easily get away with basically telling people, "If this doesn't work, your Sopel is too old and you should update it." 😁